### ISSUE TITLE
Seed encryption and plausible deniability

### ISSUE THREAD

**Issue by tevador** [tevador/polyseed#13]
The seed encryption feature currently sets a bit flag in the mnemonic phrase that marks the phrase as being encrypted. This is done for UX reasons so users don't accidentally try to restore the encrypted polyseed without a password, which does happen with the 25-word mnemonic phrase and causes confusion when users find an empty wallet.

However, when decrypting a polyseed, there is no way to check if the password that was entered is correct. Polyseed will simply XOR the secret seed with whatever bit mask is generated from the password. This allows users to create a second "duress wallet" by decrypting the polyseed with a different password than the one that was used to encrypt it.

However, some people have expressed concern that the presence of the "encrypted" bit-flag is bad for plausible deniability because an attacker might suspect trickery from the mere fact that the mnemonic seed is encrypted.

I think the current functionality provides sufficient plausible deniability even if it does not hide the fact that the seed is encrypted. However, I'm opening this discussion to hear more opinions.

**Comment by chaserene** [comment]
I can't really think of an in-between here, and if it comes down to that (better UX vs better personal security), I'm in favor of removing the "encrypted" bit flag to allow for perfect plausible deniability.

for reference, [BIP-39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#from-mnemonic-to-seed) works along the same trade-off. some wallets that rely on this passphrase functionality ask for the passphrase by default, check if the wallet is without any past activity, and if so, ask you to confirm the passphrase. this could be used as a UX crutch.

it could be even improved with specific suggestions (remind you of the passphrase option if you didn't enter any, or if you did, remind you that maybe you didn't set a passphrase initially). I'm not sure about this, but it wouldn't appear for anyone who has prepared a "duress wallet".

**Comment by tevador** [comment]
With the current functionality, the most paranoid users can have an unencrypted phrase and 2 passwords that will transform the phrase into the actual wallet (first encrypt with the first password, then decrypt with the second one).

> I'm in favor of removing the "encrypted" bit flag

It's not that simple. There are already Polyseed integrations in the wild, so we would need a backwards compatible solution.

> check if the wallet is without any past activity, and if so, ask you to confirm the passphrase

Why can't the attacker do the same? Essentially, everyone using an encrypted seed would need to put some funds into the "decoy wallet", which would negate this UX crutch.

> remind you that *maybe* you didn't set a passphrase initially

IMO, this would only cause more confusion.

> it wouldn't appear for anyone who has prepared a "duress wallet"

How would the software know? And if the software knows, why can't the attacker find out the same way that you have a decoy wallet?

**Comment by chaserene** [comment]
> the most paranoid users can have an unencrypted phrase and 2 passwords

did you mean encrypted here? if yes, let me illustrate the situation: if an attacker that can access the mnemonic can tell whether it's encrypted or not, they can be more efficient and take less risks. this is because they can focus on victims with encrypted mnemonics and expend more energy on pressing them. if crypto wrench attacks become part of the public's understanding (which they will), unencrypted mnemonics will become a Schelling point, and the encrypted minority will be in even greater danger.

contrast this with a standard that where attackers can't tell if it's encrypted or not. the average expected value and thus the risk-adjusted return of pressing a victim becomes lower because they have to engage with more of them. if they still decide to do it, it will be harder to correctly call bluffs because non-technical (=less likely to have a "duress wallet") and technical (=more likely) victims will form a single "victim pool".

I heavily oppose the labeling of the preference for personal security as "paranoia". just like privacy, security is for everyone.

> It's not that simple. There are already Polyseed integrations in the wild, so we would need a backwards compatible solution.

isn't that an unavoidable burden if we change anything in this aspect? each new major version carries the risk of additional obsoleted standards and required compatibility features.

> Why can't the attacker do the same?

this crutch targets the case you outlined first, when a user tries to restore a wallet with past activity but forgot about the password (or made a typo), and is surprised to find an untouched wallet. as you said, for duress situations, you ideally have a "duress wallet" that has activity (which doesn't trigger the crutch).

> How would the software know? 

again, this presumes you already have a "duress wallet" prepared. my arguments here are for enabling perfect plausible deniability. most people who want to take advantage of that will do the required preparations.

---

another con: with the unencrypted/encrypted distinction, IIUC, once you create a non-encrypted mnemonic, you can't generate new password-derived wallets, nor for/in duress situations, neither for other reasons. you have to create a new mnemonic, recreate backups, adjust everything else that relies on that wallet, and move funds (which most people will do with one huge multi-input transaction, which is horrible for privacy). I consider this a major setback.

**Comment by tevador** [comment]
> did you mean encrypted here?

I meant what I wrote. You can have a technically encrypted Polyseed that looks unencrypted if you use 2 passwords.

> isn't that an unavoidable burden if we change anything in this aspect?

It's avoidable if we don't make the breaking change you are proposing.

Let's wait for more feedback.

**Comment by rbrunner7** [comment]
I made a systematic review of Monero wallet apps to see whether they support Polyseed, and if yes to what degree, with special attention to the use of encrypted Polyseeds.

TL;DR: The "official" wallet apps are the last remaining dedicated Monero wallet apps that don't support Polyseed, which IMHO is a bit of a shame, and we have unfortunate interoperability problems with encrypted Polyseeds that we probably should try to solve.

---------------------

The pleasing result of my review is that most wallet apps have at least basic Polyseed support now in Spring 2026. There are two groups of exceptions: multi-coin wallets like Edge Wallet and ZelCore on the one side, and the "official" Monero CLI wallet app, GUI wallet app and Wallet RPC server on the other side.

In the following I use the following terms:

*encryption* or *encrypted* means using the Polyseed feature of encrypting a seed, in the sense of the library method [polyseed_crypt](https://github.com/tevador/polyseed/blob/dfb05d8edb682b0e8f743b1b70c9131712ff4157/src/polyseed.c#L385) and the [is_encrypted](https://github.com/tevador/polyseed/blob/dfb05d8edb682b0e8f743b1b70c9131712ff4157/src/features.h#L27) feature bit query. Thanks to the feature bit, a Polyseed knows whether it's encrypted or not.

*seed offset* means using a passphrase to modify a seed by "subtracting" the passphrase, or more exactly a hash of it, to arrive at the seed to actually use, in the sense of the [decrypt_key](https://github.com/monero-project/monero/blob/ea9be68fb66f66605d857e1ee13e92fb1d50423b/src/cryptonote_basic/cryptonote_format_utils.cpp#L1666) function in the Monero core codebase. A standard Monero 25 word seed, in the following called *legacy seed*, does not contain any info whether you have to apply a seed offset passphrase or not for accessing the "true" wallet.

Using a seed offset with a private key specified through a Polyseed may not follow the "spirit" of Polyseed because in a way it duplicates its native encryption functionality, but technically it's no problem of course, and you you can do seed offsets with Polyseeds and legacy seeds symmetrically, while encryption is Polyseed-exclusive. Plus, as argued in this thread, plausible deniability can be seen as better with seed offsets.

Regarding Polyseed support, use of encryption and seed offsets we can group the wallet apps as follows:

A) Apps that don't support Polyseed in any way. Example: Monero GUI wallet
B) Apps that only offer basic Polyseed support, without encryption nor seed offsets. Example: Skylight Wallet 1.0.8
C) Apps that support Polyseed including seed offsets, but not encryption. Example: Feather Wallet
D) Apps that support Polyseed including seed offsets and encryption. Sole example as of now: Cake Wallet

No wallet app lets you encrypt a Polyseed *and* also use a seed offset, thus we can neglect this combination, which does not make much sense anyway.

Let's look in detail at what the single wallet app in group D, Cake wallet, does:

At wallet creation time you can tell it which kind of seed to use. If you go for Polyseed, and you specify the optional passphrase, the app will produce a properly encrypted Polyseed, with the feature bit set. At least in version 5.9.0 there is no option to prevent encryption and let it use the passphrase as a seed offset - if you give it a passphrase, the result will be an encrypted Polyseed.

When restoring a wallet from seed, it supports both encryption and seed offsets. If you give a Polyseed and specify the optional passphrase, it checks the feature bit. If that says "encrypted", it decrypts the seed, if not, it uses the passphrase as a seed offset. Net result of this very simple algorithm is that Cake Wallet can restore wallets using any Polyseed.

Unfortunately we currently have a situation where the probably most used and thus most important smartphone wallet app produces encrypted Polyseeds which no other wallet app can use to properly restore the same wallet.

As a member of group C) Feather Wallet at least rejects encrypted Polyseeds with an "unsupported" error message. Other apps in group C) and all that I could find in group B) simply ignore the feature bit, with the very unfortunate result that they take the encrypted bits as the private key to use and restore an altogether different wallet without warning, which is dangerous and can lead to confusion.

Sometimes, if you have to use a seed, that's your only chance to recover funds, and IMHO the very last thing you want to encounter as a user in such a situation is an interoperability problem like it can occur here.

**Comment by j-berman** [comment]
FWIW, I agree with @rbrunner7 's opinions shared in the NWLB channel given the current state of Polyseed support by wallets in the ecosystem.

Specifically:

> I have a strong opinion that for restore all wallets should standardize on Cake Wallet's method, that can restore everything, except combinations that don't make sense

(Where as you noted, combinations that don't make sense = wallets that support both a seed offset and a passphrase)

And:
> What they do on wallet create is another story, and I wouldn't mind author's personal preference influencing that

My view is it would make sense to update the Polyseed spec to include the seed offset mechanism currently used by Feather and Cake as well. And the spec can indicate that wallet designers can choose to support either the seed offset or the encryption feature on wallet creation (the benefit of the seed offset is plausible deniability, the benefit of an encrypted seed is a stronger UX protection that lets the user know they used a passphrase).

By including the seed offset in the spec, it would lend to wallets across the ecosystem knowing how to restore polyseeds created by other wallets.

______

> Other apps in group C) and all that I could find in group B) simply ignore the feature bit, with the very unfortunate result that they take the encrypted bits as the private key to use and restore an altogether different wallet without warning, which is dangerous and can lead to confusion.

This is very unfortunate, and in my view is a problem these wallets should solve on their end. A suggested solution could be:

If the wallet has *already* imported polyseeds that have the encryption bit set, but the private key used is the encrypted bits, then the wallets can pop a flashing warning on first load that explains that a bug in the software incorrectly imported the wrong private key from the given polyseed. The wallet can indicate that the polyseed the wallet has actually been using is <seed> (the same seed but with encrypted feature bit set to 0), and tell the user to write that seed down.

That solution could work out alright because presumably these wallets would have gotten bug reports if any of their users actually did try to restore an encrypted polyseed in the wallet, and their funds didn't show up. Hence, likely the case that these wallets have very few users affected by the issue.

**Comment by tevador** [comment]
> *seed offset* means using a passphrase to modify a seed by "subtracting" the passphrase, or more exactly a hash of it, to arrive at the seed to actually use, in the sense of the [decrypt_key](https://github.com/monero-project/monero/blob/ea9be68fb66f66605d857e1ee13e92fb1d50423b/src/cryptonote_basic/cryptonote_format_utils.cpp#L1666) function in the Monero core codebase. A standard Monero 25 word seed, in the following called legacy seed, does not contain any info whether you have to apply a seed offset passphrase or not for accessing the "true" wallet.

The "subtracting" of the seed offset is done modulo the size of the prime order subgroup of Curve25519, which makes sense if the Polyseed key is used as a Curve25519 private key.

However, both Carrot and Jamtis use the Polyseed key as a symmetric key, not an elliptic curve scalar, so it's unclear what "seed offset" should do in that case.

**Comment by rbrunner7** [comment]
> However, both Carrot and Jamtis use the Polyseed key as a symmetric key, not an elliptic curve scalar, so it's unclear what "seed offset" should do in that case.

Not sure I fully understand. Does that mean that "seed offsetting" in general needs an adjustment for Carrot, regardless of whether a legacy seed or a Polyseed gets used?

**Comment by tevador** [comment]
> Does that mean that "seed offsetting" in general needs an adjustment for Carrot

Yes, it needs to be properly specified for symmetric secret keys. Treating the Carrot `s_m` secret as a Curve25519 scalar for the purpose of the seed offset is one option (it would lose 4 bits of entropy in the process). An arguably better option would be to simply XOR with a 256-bit offset.

> regardless of whether a legacy seed or a Polyseed gets used?

I wasn't aware that having a legacy seed was even an option for Carrot wallets. I would advise against supporting legacy seeds with new wallet types, unless we think it's OK to require users to manually input their hierarchy type (legacy/Carrot/Jamtis) when restoring a wallet (this would come with its own set of security issues, not just the UX).
